source/vir/src/split_expression.rs

```diff
@@ -0,0 +1,530 @@
+                    BndX::Choose(binders, _, _) => binders.iter().map(|b| b.name.clone()).collect(),
+                    BndX::Lambda(binders) => binders.iter().map(|b| b.name.clone()).collect(),
+                };
+                !binded_names.iter().any(|bname| **bname == **name)
+            });
+            let arg_map = arg_map;
+            let e1_replaced = subsitute_argument(e1, &arg_map)?;
+            let bnd_replaced: Bnd = match &bnd.x {
+                BndX::Let(binders) => {
+                    let mut new_binders: Vec<Binder<Exp>> = vec![];
+                    for old_b in &**binders {
+                        let new_b = BinderX {
+                            name: old_b.name.clone(),
+                            a: subsitute_argument(&old_b.a, &arg_map)?,
+                        };
+                        new_binders.push(Arc::new(new_b));
+                    }
+                    Spanned::new(bnd.span.clone(), BndX::Let(Arc::new(new_binders)))
+                }
+                BndX::Quant(quant, bndrs, _trigs) => Spanned::new(
+                    bnd.span.clone(),
+                    BndX::Quant(quant.clone(), bndrs.clone(), Arc::new(vec![])),
+                ),
+                _ => {
+                    return Err((
+                        exp.span.clone(),
+                        format!("Unsupported binder during subsitution"),
+                    ));
+                }
+            };
+            let new_exp = ExpX::Bind(bnd_replaced, e1_replaced);
+            SpannedTyped::new(&exp.span, &exp.typ, new_exp)
+        }
+        ExpX::Const(_) => exp.clone(),
+        _ => {
+            return Err((exp.span.clone(), format!("Unsupported expression during subsitution")));
+        }
+    };
+    Ok(result)
+}
+
+fn is_bool_type(t: &Typ) -> bool {
+    match &**t {
+        TypX::Bool => true,
+        TypX::Boxed(b) => is_bool_type(b),
+        _ => false,
+    }
+}
+
+pub(crate) fn tr_inline_expression(
+    body_exp: &Exp,
+    params: &Params,
+    exps: &Exps,
+) -> Result<Exp, (Span, String)> {
+    match *body_exp.typ {
+        TypX::Bool => (),
+        _ => {
+            return Err((
+                body_exp.span.clone(),
+                "Note: skip inlining due to verifier(external_body)".to_string(),
+            ));
+        }
+    };
+    let mut arg_map: HashMap<Arc<String>, Exp> = HashMap::new();
+    let mut count = 0;
+    for param in &**params {
+        let exp_to_insert = &exps[count];
+        if !crate::ast_util::types_equal(&param.x.typ, &exp_to_insert.typ) {
+            panic!("Internal error: arg type mismatch during expression inlining");
+        }
+        arg_map.insert(param.x.name.clone(), exp_to_insert.clone());
+        count = count + 1;
+    }
+    return subsitute_argument(body_exp, &arg_map);
+}
+
+pub(crate) fn pure_ast_expression_to_sst(ctx: &Ctx, body: &Expr, params: &Params) -> Exp {
```

Comment:
Rather than running expr_to_exp again, you can get the already computed SST body from `fun_ssts`.

Reply:
I will try to use the information in `fun_ssts`. Thanks!